### DISCUSSION TITLE
!!Beta Testing Needed!! Please help us out!

### DISCUSSION THREAD

**Discussion by danner26** [netbox-community/Device-Type-Library-Import#80] [Announcements]
**Update**: This beta has been officially released. Please see (release v2.0.0)[https://github.com/netbox-community/Device-Type-Library-Import/releases/tag/v2.0.0-2023-03-24] for more information. Also, thank you to everyone who assisted with this releases testing!

Hello all,

Tonight I have released the first iteration of the new codebase for this repository. I have completely rewritten the application, optimizing function calls and loops as much as possible, moved logging to its own callable section with verbosity options, abstracted the function calls into 3 separate classes, and more. This is a huge overhaul, and I have done my due diligence before releasing the beta. Now I need your help!

Please go ahead and clone the new branch [pre-release-newbase](https://github.com/netbox-community/Device-Type-Library-Import/tree/pre-release-newbase) which contains the current up to date copy of the new baseline. Please test it out, import devices and modules and assist me with finding any regressions or bugs.

I will keep this beta open for 2 weeks, so on March 23, 2023 we will be merging this to Master unless we have too many bugs/regressions to fix before. As of today, March 9th, 2023 we are on schedule for release.

Thank you for all who contribute!

**Comment by danner26** [comment]
Calling all past contributors, please if you can test out the new beta release before it is merged to master. This is a big change to the codebase so we need as many beta testers as possible. Thank you!

@dmcken 
@ndom91 
@mtinberg 
@k0rventen 
@dalrrard 
@FloEisen 
@gaby 
@p-rintz 
@rmundel 
@suom1 
@rsp2k 
@eN0Rm 
@lordwolf2004 
@FlorianHeigl 
@mayafox 
@minitriga

**Reply by dmcken** [reply]
Glad to hear it, I'll test in the morning.

**Reply by danner26** [reply]
> Glad to hear it, I'll test in the morning.

Thank you David!

**Comment by p-rintz** [comment]
Some notes:
```
/opt/Netbox-Device-Type-Library-Import/repo.py:78: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if folder.casefold() is not "testing":
```

You will need to handle the case where you run into the second except in the create functions in the netbox_api.py (ie. add a 'continue' after line 92 in netbox_api.py, but the same would likely need to happen elsewhere in the api, though I havent run into it). Right now it will run into the following issue where the api object was never created successfully:
```
{"weight":["A valid number is required."]}
Traceback (most recent call last):
  File "/opt/Netbox-Device-Type-Library-Import/nb-dt-import.py", line 53, in <module>
    main()
  File "/opt/Netbox-Device-Type-Library-Import/nb-dt-import.py", line 25, in main
    netbox.create_device_types(device_types)
  File "/opt/Netbox-Device-Type-Library-Import/netbox_api.py", line 95, in create_device_types
    self.device_types.create_interfaces(device_type["interfaces"], dt.id)
UnboundLocalError: local variable 'dt' referenced before assignment
```

That said, setting weight seems to be broken.

Netbox API seems to have changed:
<span><span><span class="model"><span class=""><span class="inner-object">

```
weight | number($decimal)title: Weightx-nullable: true
-- | --
weight_unit | stringtitle: Weight unitEnum:
```

Instead of :
```
 - value: 31
   unit: kg
```

The capitalization of the Vendor 'neousys' will throw the following error (error is obvious, no idea how you have changed this case before, where the name changes, but the slug stays the same):
`Error during manufacturer creation. - [{"slug":["manufacturer with this slug already exists."]},{"slug":["manufacturer with this slug already exists."]},{}]`

(same for 'D-link')

Some others have changed their name, but also changed slug: ie. TrippLite (keeping the old manufacturers in the netbox db)

Similar issue with device types where name has changed, but slug hasnt:
```
{'manufacturer': {'name': 'Cisco', 'slug': 'cisco'}, 'model': 'Catalyst 9130AXI-B', 'slug': 'c9130axi-b', 'part_number': 'C9130AXI-B', 'u_height': 0, 'is_full_depth': False, 'interfaces': [{'name': 'WiFi', 'type': 'ieee802.11ax'}, {'name': 'GigabitEthernet0', 'type': '5gbase-t'}], 'console-ports': [{'name': 'console', 'type': 'rj-45'}, {'name': 'usb', 'type': 'usb-mini-b'}]}
The request failed with code 400 Bad Request: {'__all__': ['Device type with this Manufacturer and Slug already exists.']}
```

The following fails as well, as the parent subdevice_role was added after the device was created. And its not being updated.
`Error '[{"__all__":["Subdevice role of device type (RS1619xs+) must be set to \"parent\" to allow device bays."]}]' creating Device Bay`

This is better situated in the DeviceTypeLibrary itself, but the following wasnt correctly migrated to the new Name/Slug:
https://github.com/netbox-community/devicetype-library/blob/master/module-types/TrippLite/SNMPWEBCARD.yaml

**Reply by danner26** [reply]
Hey, thank you for the in-depth review. I am going to try to tackle each points, please do let me know if I miss something.. So anyway, here we go:

1. In the DeviceType library there were multiple issues which have been resolved as of last night. The first being the `neousys` vendor. Whoever created these files incorrectly created the folder/files and this was not caught previously. I have since fixed the issue in the upstream DeviceType-Library, but unfortunately there is no current way to fix existing instances in NetBox without some manual intervention. Along with the `neousys` issue in the upstream library, the following vendors also received fixes last night: Avocent (previously files had `manufacturer: Cyclades`), D-Link (folder name correction), Neousys (folder name correction, but again you will need to remove/manually fix the vendor if you have already created it), Tripp Lite (another vendor you will need to delete/manually correct if previously imported), and Zyxel.
Please see [this PR](https://github.com/netbox-community/devicetype-library/pull/1185) for more information

2. The DeviceType-Library is a complete mess, and over the past 2 months we have been attempting to standardize the massive amount of data within it. Unfortunately it is not in scope with the release to solve the data mismatch issues, since ultimately that is going to have to be a manual correction on the users end (either by deleting the previously imported device/module or making the manual correction itself). In fact, this repo does not check for non-defined interfaces and handle those, it does not update existing interfaces properly, etc but those are in the future plans.. Just not within this release

3. I believe I corrected the `RS1619xs+` last night in this [DeviceType-Library PR](https://github.com/netbox-community/devicetype-library/pull/1185/files#diff-751c7a8aab06e8f4abbf85d45cdf4bc10222f10339b82a6cb02e9648d52afd21). Can you use `--verbose` and see if it is pulling the latest copy of the repo for you?

4. In regards to the weight issue, I created a bug report in the pynetbox repository, as I thought the NetBox API/model was in line with what we have here. Do you have an example of how you think we should model weight in the DeviceType-Library that would work? If so, then I would be more than happy to make the necessary corrections in the DeviceType-Library. The way I was reading the stack trace last night the pynetbox library wasn't properly passing the data to the NetBox API, but I could be completely wrong.. I appreciate any insight you may have on this!

5. I thought I fixed the first one before I merged the code into this branch, I guess I did not. In regards to `/opt/Netbox-Device-Type-Library-Import/repo.py:78: SyntaxWarning: "is not" with a literal. Did you mean "!="? if folder.casefold() is not "testing":` could you please submit an issue for me?

Seriously, thank you so much for the in-depth, very thorough review! This is extremely helpful and I appreciate your continued assistance!

**Reply by p-rintz** [reply]
The issue with 3. is that the device already exists in my library and isnt being updated automatically to have the parent subdevice_role. It does have the updated 'repo'.

Just something I found while looking into things.

For 4.:

Based on the documentation, I think it should be: (quick test confirms this works)
```
weight: 31
weight_unit: kg
```

As a side note: it might be a good idea to have a cli toggle or something to clean up old non-used device types, to assist in cleaning up these kind of issues in already imported netbox databases.

It might also be good to make it clearer which device-types have these issues. 
For my debugging I inserted some prints into the try's to output which device type/manufacturer exactly ran into problems.

**Reply by danner26** [reply]
> The issue with 3. is that the device already exists in my library and isnt being updated automatically to have the parent subdevice_role. It does have the updated 'repo'.

Ah okay, that makes more sense. Unfortunately right now there isnt a way for us to update the already created interfaces/devices. I hope to tackle this in a future release but for now it just required manual intervention either by you deleting the device/module in netbox, then rerunning or making the correction by hand

> Based on the documentation, I think it should be:

I will give that a shot when I get back to the office! If this is indeed the fix, I will correct this in the DeviceType-Library repository.. Hopefully it is that simple

> As a side note:

Do you mean a toggle that would ultimately look for unused devices/modules and remove them before the creation process? If that is what you mean, how do we stop this from deleting manually created devices/modules that the user wants, but are not in the library?

> It might also be good to make it clearer which device-types have these issues.

I agree. Could you create an issue/feature request to add the device/module name to the error output when an import fails? I would appreciate that (and if you have already working code for this feel free to submit a PR against the branch!)

**Reply by p-rintz** [reply]
I edited my reply. The weight change does work and it now imports the device type with this.

```
{'manufacturer': {'name': 'Inspur', 'slug': 'inspur'}, 'model': 'NF5180M6', 'slug': 'inspur_nf5180m6', 'part_number': 'NF5180M6', 'u_height': 1, 'is_full_depth': True, 'airflow': 'front-to-rear', 'comments': '[Inspur NF5180M6 Technical Specifications](https://en.inspur.com/en/servers/rack_servers/2552517/index.html)', 'weight': 31, 'weight_unit': 'kg', 'console-ports': [{'name': 'BMC/System Serial', 'type': 'other'}], 'module-bays': [{'name': 'PSU0', 'position': 'PSU0'}, {'name': 'PSU1', 'position': 'PSU1'}, {'name': 'OCP3', 'position': 'OCP3'}, {'name': 'PCIe0', 'position': 'PCIe0', 'label': 'Left PCIe Riser-Card'}, {'name': 'PCIe1', 'position': 'PCIe1', 'label': 'Left PCIe Riser-Card'}, {'name': 'PCIe2', 'position': 'PCIe2', 'label': 'Right PCIe Riser-Card'}], 'interfaces': [{'name': 'LAN0', 'type': '10gbase-x-sfpp'}, {'name': 'LAN1', 'type': '10gbase-x-sfpp'}, {'name': 'BMC', 'type': '1000base-t', 'mgmt_only': True}]}
Device Type Created: Inspur - NF5180M6 - 2322
```

**Reply by p-rintz** [reply]
> Do you mean a toggle that would ultimately look for unused devices/modules and remove them before the creation process? If that is what you mean, how do we stop this from deleting manually created devices/modules that the user wants, but are not in the library?

For my usecase, it wouldnt need to. It would need to be clear on this. If you dont want to have this, that is completely fine. Ill just have to create a small script myself. Im just not looking forward to deleting 2000+ device types by hand to correct old, (potentially) incorrect data 😆

**Reply by danner26** [reply]
> I edited my reply. The weight change does work and it now imports the device type with this.

Awesome! I will do my due diligence with this and then create a PR to fix it.

> For my usecase, it wouldnt need to.

I completely understand, I suggest you create your own script for the time being. I am thinking we pass on this for the time being to reduce the potential increase in complexity, especially since this is already complex enough. Like I had mentioned before I would like to investigate having logic to update existing devices and what not, so maybe this is something we can look at after this pre-release is released.

**Reply by p-rintz** [reply]
I created a PR in the DL itself for the SNMPWEBCARD issue and here for the other two small issues we discussed.

**Reply by danner26** [reply]
Awesome, merged! I am waiting for the demo to be reset to continue the testing

**Reply by danner26** [reply]
Perfect that worked! I will start updating weight in the devicetype-library

**Reply by danner26** [reply]
@p-rintz take a look at [this pr](https://github.com/netbox-community/devicetype-library/pull/1187)

It fixes the weight issue

**Reply by danner26** [reply]
Sweet, I just tested the changes and now all of the weight objects that were failing previously are working! Thank you for your assist on that

**Comment by dmcken** [comment]
I've done a complete import against a new instance of 3.4.5 and at least from a numbers and spot check it looks to have imported fine:

```
Package Installed https://github.com/netbox-community/devicetype-library.git
103 Vendors Found
2208 Device-Types Found
Modules Enabled. Creating Modules...
30 Module Vendors Found
287 Module-Types Found
Error '{"__all__":["Module type with this Manufacturer and Model already exists."]}' creating module type: {'manufacturer': {'name': 'Allied Telesis', 'slug': 'allied-telesis'}, 'model': 'SBx81GS24a', 'part_number': 'AT-SBx81GS24a', 'interfaces': [{'name': 'port1.{module}.1', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.2', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.3', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.4', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.5', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.6', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.7', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.8', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.9', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.10', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.11', 'type': '1000base-x-sfp'}, {'name': 'port1.{module}.12', 'type': '1000base-x-sfp'}]}
---
2208 devices created
56507 interfaces/ports updated
1 manufacturers created

real	52m2.124s
user	1m49.930s
sys	0m11.973s
```

You can ignore the error, I've already raised a PR 1209 to remove that module.

**Reply by danner26** [reply]
Awesome! Thank you for your assistance :)

**Reply by dmcken** [reply]
Redone locally on my home PC on NVME storage (behind virtualbox) on netbox 3.4.6. The PR was accepted, so error is gone.
```
Package devicetype-library is already installed, updating /home/dmcken/Device-Type-Library-Import/repo
103 Vendors Found
2210 Device-Types Found
Modules Enabled. Creating Modules...
30 Module Vendors Found
286 Module-Types Found
---
2210 devices created
56518 interfaces/ports updated
104 manufacturers created

real    10m42.798s
user    0m41.650s
sys     0m2.123s
```

I've looked through some of the device types I've added for Juniper and Mikrotik to the library and they are correct.

A 2nd run (the device-types have already been added), comes back with the following:
```
Package devicetype-library is already installed, updating /home/dmcken/Device-Type-Library-Import/repo
103 Vendors Found
2210 Device-Types Found
Modules Enabled. Creating Modules...
30 Module Vendors Found
286 Module-Types Found
---
0 devices created
0 interfaces/ports updated
0 manufacturers created

real    2m49.525s
user    0m27.376s
sys     0m0.818s
```

I guess if its easy to add I would put a line to display the number of module-types added (I know I noticed when I was performing my updates but I forget why I didn't add it to that pull request).

**Reply by danner26** [reply]
Sounds like a good addition that is probably actually a regression from the original.. I might have missed that part. If you could, please create an issue so that I can track the addition of the modules created line
